data/secscan_model/secscan_v4_model.py

```diff
@@ -0,0 +1,152 @@
+)
+
+
+logger = logging.getLogger(__name__)
+
+
+class ScanToken(namedtuple("NextScanToken", ["min_id"])):
+    """
+    ScanToken represents an opaque token that can be passed between runs of the security worker
+    to continue scanning whereever the previous run left off. Note that the data of the token is
+    *opaque* to the security worker, and the security worker should *not* pull any data out or modify
+    the token in any way.
+    """
+
+
+class V4SecurityScanner(SecurityScannerInterface):
+    """
+    Implementation of the security scanner interface for Clair V4 API-compatible implementations.
+    """
+
+    def __init__(self, app, instance_keys, storage):
+        self.app = app
+
+        validator = SecurityConfigValidator(
+            app.config.get("FEATURE_SECURITY_SCANNER", False),
+            app.config.get("SECURITY_SCANNER_ENDPOINT"),
+        )
+        if not validator.valid():
+            logger.warning("Failed to validate security scanner V2 configuration")
+            return
+
+    def load_security_information(self, manifest_or_legacy_image, include_vulnerabilities=False):
+        if not isinstance(manifest_or_legacy_image, datatypes.Manifest):
+            return None
+
+        status = None
+        try:
+            status = ManifestSecurityStatus.get(manifest=manifest_or_legacy_image._db_id)
+        except ManifestSecurityStatus.DoesNotExist:
+            return SecurityInformationLookupResult.with_status(ScanLookupStatus.NOT_YET_INDEXED)
+
+        if status.index_status == IndexStatus.FAILED:
+            return SecurityInformationLookupResult.with_status(ScanLookupStatus.FAILED_TO_INDEX)
+        if status.index_status == IndexStatus.UNSUPPORTED:
+            return SecurityInformationLookupResult.with_status(
+                ScanLookupStatus.UNSUPPORTED_FOR_INDEXING
+            )
+        if status.index_status in [
+            IndexStatus.TIMED_OUT,
+            IndexStatus.WAITING,
+            IndexStatus.IN_PROGRESS,
+        ]:
+            return SecurityInformationLookupResult.with_status(ScanLookupStatus.NOT_YET_INDEXED)
+
+        assert status.index_status == IndexStatus.COMPLETED
+
+        # TODO(alecmerdler): Fetch data from Clair API
+        return SecurityInformationLookupResult.with_status(ScanLookupStatus.NOT_YET_INDEXED)
+
+    def perform_indexing(self, start_token=None):
+        whitelisted_namespaces = self.app.config.get("SECURITY_SCANNER_V4_INDEXING_WHITELIST", [])
+
+        def eligible_manifests(base_query):
+            return (
+                base_query.join(Repository)
+                .join(User)
+                .where(
+                    User.username << whitelisted_namespaces if len(whitelisted_namespaces) else True
+                )
+            )
+
+        def id_for(id_fn):
+            return eligible_manifests(Manifest.select(id_fn, Repository, User)).scalar()
+
+        min_id = id_for(fn.Min(Manifest.id))
+        max_id = id_for(fn.Max(Manifest.id))
```

Comment:
This `max` is going to be very, very, very slow.

Reply:
Same with the `min`, I assume then. Should we remove the namespace filtering logic here (since it is already in the `batch_query` argument to `yield_random_entries()`) and just use:
```python
min_id = Manifest.select(fn.Min(Manifest.id)).scalar()
max_id = Manifest.select(fn.Max(Manifest.id)).scalar()
```
Would this work?

Reply:
`min` might be a bit faster, but yeah, we're going to filter it anyway.